### PR TITLE
Nested 'only' parameter

### LINKED ISSUE #402: Dynamically specify 'only' parameter for nested fields
I'm able to specify the `only` parameter when instantiating a schema (`MySchema(only=('field1', 'field2')`), however, this does not work for nested fields. I'd like to be able to have something like `MySchema(only=('field1', ('field2', ('subfield1', 'subfield3'))))` where _field2_ is a nested field of which only _subfield1_ and _subfield2_ should be serialized.

I'd be willing to implement it, however, first I wanna know if it has a chance to get into marshmallow. Secondly, it seems to not be a straight forward addition; instead of (or additionally to) during instantiation of the nested schema, the _only_ parameter needs to be passed when serializing. So I'd appreciate any hints of how to implement this properly.

The motivation behind this are APIs like [JSONAPI](http://jsonapi.org/) which allow to dynamically specify fields like so: `GET /articles?include=author&fields[articles]=title,body&fields[people]=name`. One could argue that resources are not really rendered in a nested fashion in JSONAPI responses. So lets assume, we use a (not yet implemented) `IncludeSchema` instead. That will, however, leave us with the same challenges of specifying the _only_ parameter dynamically.

### PR DESCRIPTION
**Pull request by Archelyst**
This allows to specify a nested `only` parameter both at schema instantiation time and during dump() like so:

``` python
class ChildSchema(Schema):
    foo = fields.Field()
    bar = fields.Field()
    baz = fields.Field()

class ParentSchema(Schema):
    bla = fields.Field()
    bli = fields.Field()
    blubb = fields.Nested(ChildSchema)

data = dict(bla=1, bli=2, blubb=dict(foo=42, bar=24, baz=242))

# either when instantiating
sch = ParentSchema(only=('bla', ('blubb', ('foo', 'bar'))))
result = sch.dump(data)

#or when dumping
sch = ParentSchema()
result = sch.dump(data, only=('bla', ('blubb', ('foo', 'bar'))))
```

It is fully backwards compatible. Fixes #402

### CODE REVIEW

**Comment by Archelyst** [comment]
Concerning the syntax: `only=('bla', ('blubb', ('foo', 'bar')))` means "only bla and blubb and of blubb only foo and bar". If anybody has a better idea, it's probably not hard to change it.

**Comment by sloria** [comment]
Thanks for the PR! This is a nice feature. I will review this soon.

**Comment by Archelyst** [comment]
Any thoughts about this one?

File: marshmallow/schema.py [hunk 1]
```diff
@@ -452,7 +452,7 @@ def accessor(cls, func):
 
     ##### Serialization/Deserialization API #####
 
-    def dump(self, obj, many=None, update_fields=True, **kwargs):
+    def dump(self, obj, many=None, update_fields=True, only=None, **kwargs):
```
Comment (sloria) [inline marshmallow/schema.py]: I'm not sure about adding `only` as a param here. I think it'd be good to limit the number of params to `dump` and `load`. Otherwise, users might expect all `Meta` options (e.g. `only`, `exclude`, `include`) to be parameters.

**Comment by sloria** [comment]
How well does this approach generalize to `exclude`? This feature was requested in https://github.com/marshmallow-code/marshmallow/issues/151 .

File: marshmallow/schema.py [hunk 2]
```diff
@@ -668,6 +686,11 @@ def _update_fields(self, obj=None, many=False):
         if excludes:
             field_names = field_names - excludes
         ret = self.__filter_fields(field_names, obj, many=many)
+        # Set nested only parameters overriding the only parameters declared on the fields
+        for field_name, field in [(fn, f) for fn, f in ret.items() if hasattr(f, 'only')]:
+            field.only_explicit = None
+            if only_nested and field_name in only_nested:
```
Comment (sloria) [inline marshmallow/schema.py]: Minor: I think this can be simplified to `if field_name in only_nested:` , since `only_nested` will always be a `dict`.

**Comment by sloria** [comment]
Initial pass finished. :+1: Sorry for the delay on my review; things have been busy at work.

**Comment by sloria** [comment]
Also, I am bit concerned about the complexity this feature adds to the codebase, which makes me hesitant to merge this. I welcome feedback on whether this is a necessary feature.

File: marshmallow/schema.py (hunk 1 above)
Comment (Archelyst) [inline marshmallow/schema.py]: Ok, that's a good point. I'll see, if I can implement my use cases without this. What's the rationale for specifying `many` here? There are three and a half places where the schema can be configured (Meta, **init**, and dump/load); I don't see the logic yet, what goes where and why.

**Comment by Archelyst** [comment]
> How well does this approach generalize to exclude? This feature was requested in #151 .

I think we could implement the same mechanism for `exclude`, however, from what I read in #151, a nested `only` parameter would suffice. It's just the inverse, it would probably be not too hard to convert any exclude to an only.

> Initial pass finished. :+1: Sorry for the delay on my review; things have been busy at work.

Thanks for the review, man. You've put so much work into this library from which I benefit greatly. No need to apologize :)

**Comment by Archelyst** [comment]
> Also, I am bit concerned about the complexity this feature adds to the codebase, which makes me hesitant to merge this. I welcome feedback on whether this is a necessary feature.

Hm, you're right, the whole part the with figuring out which fields to render is a little tricky. However, the addition at hand is mostly encapsulated in one function. I think, "nested only" is worth the extra complexity.

You've probably seen my PR for marshmallow-jsonapi. IMO it would be big and increase adoption a lot if marshmallow would fully implement jsonapi (as much as serialization is concerned). One part of that is choosing the (nested) fields dynamically. You don't happen to come to Europe any time soon? Would be awesome to chat about how we could approach this. Maybe a little hackaton? :)

**Comment by dscerri** [comment]
Thanks for all your hard work, this library has been of great benefit to me!

I agree that this adds some complexity to the codebase, however I think this feature would be very useful. My current solution involves having two parent schemas which only differ in the included fields in the nested schema, and this adds unnecessary complexity. 

An alternative dot notation was proposed in a previous PR (#180), which was maybe a little cleaner.

**Comment by Archelyst** [comment]
It's probably not too hard to create a little helper that transforms any (dot) syntax into the easier to process list syntax. However, I don't really like the Syntax in PR #180, because adding several nested fields of one nested item is very verbose: `'first_name, last_name, address.street, address.zip_code, address.city'`. Could rather be `'first_name, last_name, address.fields(street, zip_code, city)'`, that's how facebook does it. Anyhow, we could add parsers for both.

**Comment by Archelyst** [comment]
I removed the 'only' parameter from dump(). Now it doesn't look that scary any more :)

File: marshmallow/fields.py [hunk 3]
```diff
@@ -499,14 +500,14 @@ def __init__(self, cls_or_instance, **kwargs):
         if isinstance(cls_or_instance, type):
             if not issubclass(cls_or_instance, FieldABC):
                 raise ValueError('The type of the list elements '
-                                           'must be a subclass of '
-                                           'marshmallow.base.FieldABC')
+                                 'must be a subclass of '
+                                 'marshmallow.base.FieldABC')
```
Comment (deckar01) [inline marshmallow/fields.py]: Unnecessary formatting change.

Making formatting changes to code you didn't functionally modify makes a PR much more difficult to review.

I would recommend removing unnecessary formatting changes and squashing most of these commits since this PR has churned so much.

File: marshmallow/fields.py [hunk 4]
```diff
@@ -499,14 +500,14 @@ def __init__(self, cls_or_instance, **kwargs):
...
             self.container = cls_or_instance()
         else:
             if not isinstance(cls_or_instance, FieldABC):
                 raise ValueError('The instances of the list '
-                                           'elements must be of type '
-                                           'marshmallow.base.FieldABC')
+                                 'elements must be of type '
+                                 'marshmallow.base.FieldABC')
```
Comment (deckar01) [inline marshmallow/fields.py]: Unnecessary formatting change.

File: marshmallow/marshalling.py [hunk 5]
```diff
@@ -118,9 +118,9 @@ def serialize(self, obj, fields_dict, many=False,
         if many and obj is not None:
             self._pending = True
             ret = [self.serialize(d, fields_dict, many=False,
-                                    dict_class=dict_class, accessor=accessor,
-                                    index=idx, index_errors=index_errors)
-                    for idx, d in enumerate(obj)]
+                                  dict_class=dict_class, accessor=accessor,
+                                  index=idx, index_errors=index_errors)
+                   for idx, d in enumerate(obj)]
```
Comment (deckar01) [inline marshmallow/marshalling.py]: Unnecessary formatting change.

File: marshmallow/marshalling.py [hunk 6]
```diff
@@ -233,9 +233,9 @@ def deserialize(self, data, fields_dict, many=False, partial=False,
         if many and data is not None:
             self._pending = True
             ret = [self.deserialize(d, fields_dict, many=False,
-                        partial=partial, dict_class=dict_class,
-                        index=idx, index_errors=index_errors)
-                    for idx, d in enumerate(data)]
+                       partial=partial, dict_class=dict_class,
+                       index=idx, index_errors=index_errors)
+                   for idx, d in enumerate(data)]
```
Comment (deckar01) [inline marshmallow/marshalling.py]: Unnecessary formatting change.

File: marshmallow/schema.py [hunk 7]
```diff
@@ -318,7 +318,7 @@ class Meta:
         - ``index_errors``: If `True`, errors dictionaries will include the index
             of invalid items in a collection.
         - ``load_only``: Tuple or list of fields to exclude from serialized results.
-        - ``dump_only``: Tuple or list of fields to exclude from deserialization
+        - ``dump_only``: Tuple or list of fields to exclude from deserialization.
```
Comment (deckar01) [inline marshmallow/schema.py]: Unnecessary formatting change.

File: marshmallow/schema.py [hunk 8]
```diff
@@ -471,8 +471,8 @@ def dump(self, obj, many=None, update_fields=True, **kwargs):
         many = self.many if many is None else bool(many)
         if not many and utils.is_collection(obj) and not utils.is_keyed_tuple(obj):
             warnings.warn('Implicit collection handling is deprecated. Set '
-                            'many=True to serialize a collection.',
-                            category=DeprecationWarning)
+                          'many=True to serialize a collection.',
+                          category=DeprecationWarning)
```
Comment (deckar01) [inline marshmallow/schema.py]: Unnecessary formatting change.

File: marshmallow/fields.py [hunk 9]
```diff
@@ -938,7 +938,6 @@ def _serialize(self, value, attr, obj):
             return value.isoformat()
         except AttributeError:
             self.fail('format', input=value)
-        return value
```
Comment (deckar01) [inline marshmallow/fields.py:941]: Why was this removed? It seems unrelated to the feature.

This was returning `value` if an exception other than `AttributeError` occurred.

File: marshmallow/fields.py [hunk 10]
```diff
@@ -915,7 +916,6 @@ def _deserialize(self, value, attr, data):
         """Deserialize an ISO8601-formatted time to a :class:`datetime.time` object."""
         if not value:   # falsy values are invalid
             self.fail('invalid')
-            raise err
```
Comment (deckar01) [inline marshmallow/fields.py:918]: Why was this removed? It seems unrelated to the feature.

If `err` is not defined, then a separate issue should be created.

Comment (Archelyst) [inline marshmallow/fields.py:918]: Yeah, this could be removed in an act of bureaucracy and properness :) I'm more in favor of the [Boy Scout rule](bureaucracy). But if it helps merging this PR, sure, I can undo all cleanup. (By the way, I was not looking for things to clean up, I just stumbled over them.)

File: marshmallow/fields.py (hunk 3 above)
Comment (Archelyst) [inline marshmallow/fields.py]: I'll definitely squash the commits once everything is settled here.

File: marshmallow/fields.py (hunk 10 above)
Comment (deckar01) [inline marshmallow/fields.py:918]: I can't merge this PR, but I would encourage keeping PRs as concise as possible for the sake of the reviewer.

It is my personal preference to keep unrelated changes out of a PR, although it is generally acceptable to include necissary formatting fixes in files you modified in a separate commit.
http://stackoverflow.com/questions/2214480/committing-when-changing-source-formatting

Removing unrelated lines of code without testing the behavior is probably not a good idea. Opening a separate PR would help keep unrelated changes from getting blocked on each other.

Comment (sloria) [inline marshmallow/fields.py:918]: Good suggestions, @deckar01 . 

I definitely prefer to have PRs as concise and cohesive as possible. That said, I'm not a stickler on it, and I appreciate when people clean up the code. People are volunteering their valuable time to make these PRs, so I typically don't force people to do too much git "busywork".

**Comment by sloria** [comment]
I think we should move forward with this PR, as this feature has been requested by a number of users by now.

There are a couple things to address before merging:

1) Syntax: `dotted.notation` or `('nested', ('tuple', 'notation'))`?

@Tim-Erwin makes a fair point about boilerplate here: https://github.com/marshmallow-code/marshmallow/pull/405#issuecomment-191287474 . That said, we already use dotted notation when accessing attributes, so using dotted notation for `only` might be more consistent.

``` python
city = fields.Str(attribute='address.city')
```

I'm still up in the air about this issue.

2) Nested `exclude` parameter

If we allow for this notation with `only`, users will expect the same notation to work with `exclude`, so this should be implemented, too.

File: marshmallow/fields.py (hunk 10 above)
Comment (deckar01) [inline marshmallow/fields.py:918]: I went ahead and removed comments about formatting nits that didn't have additional discussion.

Thanks for working on this PR @Tim-Erwin.

**Comment by deckar01** [comment]
Although the dotted syntax is more verbose, it is very literal and therefore very readable. Nested tuple syntax representing a dictionary filter does not seem common and seems like it would be difficult to document clearly.

That said, I understand the desire to express a tree concisely and without duplication, so my proposal would be to bake the dotted syntax into the implementation and docs, then provide a utility method that flattens a nested tuple into its corresponding dotted property chains.

`flatPropertyChain('nested', ('tuple', ('notation')))`

would evaluate to

`('nested', 'tuple.notation')`

> Edit: I updated the example, because I had it wrong, which also made me realize it is a little confusing. It is not obvious that the nesting logic doesn't apply until the second level of tuples. There are also valid tuples that are invalid property chains. `('nested', ('tuple',))` is `('nested.tuple',)` or `('nested', 'tuple')`?

**Comment by sloria** [comment]
That's a sound approach, @deckar01. I'm hesitant to support two ways of doing the same thing, though. One compromise would be to support the dotted syntax out-of-the-box, then add an example in the docs and/or examples/ folder using a helper to use the tuple syntax.

**Comment by deckar01** [comment]
One concern that I have is the addition of `Nested.only_explicit`. I'm not completely sure why it is necessary to maintain a separate collection of only fields.

Here is my stab at an implementation: https://github.com/marshmallow-code/marshmallow/compare/dev...deckar01:dotted-nesting-support-402

Is there an issue with applying the `only` and `exclude` options to the nested schema directly?

**Comment by Archelyst** [comment]
Ok, I reverted the commit that introduced the formatting and unrelated changes. I'll make a separate PR for that. Also I squashed everything into one commit.

Concerning the syntax, the nested tuple one seems cleaner to me. It's closer to what's actually needed (a tree) and easier to construct programmatically. Also, it seems more concise compared to what I've seen in your tests. I'd go the other way around and provide a utility to transform dotted notation to tuples. However, I wouldn't strongly insist on that. The dotted notation looks fine as well.

I don't get your statement

> It is not obvious that the nesting logic doesn't apply until the second level of tuples.

Could you give an example for that? The programming is straight forward and does not distinguish between first and following levels.

**Comment by Archelyst** [comment]
> One concern that I have is the addition of Nested.only_explicit. I'm not completely sure why it is necessary to maintain a separate collection of only fields.

Well, `only` can be used during schema init as well as during `dump()`.

**Comment by deckar01** [comment]
> I don't get your statement [about the ambiguity of nested tuples.]
> Could you give an example for that?

`('a', ('b',), 'c') -> ('a', 'b', 'c')` or invalid?

`('a', ('b', ()), 'c') -> ('a', 'c')` or invalid?

`('a', ('b', 'c')) -> ('a', 'b', 'c')` or invalid?

`('a', ('b', ('c',)), ('b', ('d',))) -> ('a', 'b.c', 'b.d')` but currently seems to overwrite to `('a', 'b.d')`.

Dotted strings don't allow representing the same field in more than one way, and invalid dotted strings error naturally with the existing field name validation logic.

> Well, `only` can be used during schema init as well as during `dump()`.

There is no `only` parameter for `Schema.dump()`.

**Comment by Archelyst** [comment]
Ok, I get that there is a way to write invalid tuples.

> > Well, only can be used during schema init as well as during dump().
> 
> There is no only parameter for Schema.dump().

Sorry, that was bs. The reason for `only_explicit` is that you can specify `only` on `NestedField`s which should not be overwritten. Otherwise next time a schema is instantiated, not the `only` param specified on the `NestedField` will be active, but the last one specified by a nested `only` param during schema init.

**Comment by deckar01** [comment]
`update_fields` does seem to suggest `dump` has mutation side effects that can be disabled. It might be a good idea to add a test for this behavior.

**Comment by Archelyst** [comment]
Hm, do you think we should accept side effects in `dump`? That seems counterintuitive to me. `update_fields` is IMHO just an implementation detail (performance improvement for `NestedFields`) that's unfortunately visible from the outside.

**Comment by deckar01** [comment]
[The commit that introduced `update_fields`](https://github.com/marshmallow-code/marshmallow/commit/d39790255d00a7520dbf450a67d91e5f34d698c9) seems to prevent the same schema's fields from being updated more than once. Rather than performance, could this be required to keep recursively nested schema from cycling? It was originally undocumented, but later received [documentation](https://github.com/marshmallow-code/marshmallow/commit/90d11fc4813f6604872c0e53933cf3f07c6e39f4).

Something to keep in mind is that the `Nested` field's underlying `__schema` object [is always a schema instance](https://github.com/marshmallow-code/marshmallow/blob/dev/marshmallow/fields.py#L394) not a class, so you shouldn't have to worry about `dump()` mutating any classes, just schema instances that may have been reused to construct `Nested` fields.

**Comment by Archelyst** [comment]
`update_fields` is used in the case of `many=True` where it is sufficient to only update the fields once and use that configuration for all items.

`__schema` is an instance, but it shares the fields (instances) with any other instances. So while it is fine to modify the schema instance, we shouldn't modify fields if it's avoidable. And it's avoidable quite easily, why do you object to that?

**Comment by deckar01** [comment]
> we shouldn't modify fields if it's avoidable ... why do you object to that?

I suggest adding a test for this behavior. Technically you are still modifying the fields, but just adding a new property to avoid overriding `only`. 

> `update_fields` does seem to suggest `dump` has mutation side effects that can be disabled. It might be a good idea to add a test for this behavior.

**Comment by Archelyst** [comment]
> I suggest adding a test for this behavior.

Which behavior and what should the test test?

> Technically you are still modifying the fields, but just adding a new property to avoid overriding only.

That's true, but at least one shouldn't override settings a user sets with an operation like `dump()`.

My favorite solution would be to distinguish between fields on a schema class and field instances on a schema instance.

**Comment by deckar01** [comment]
I would test the purpose of `only_explicit` by dumping then asserting that fields reused on another schema have not mutated unexpectedly, and the other schema's dump is not polluted by the first.

It seems that if a `only_explicit` property is not overwritten due to a nested `only` in a second dump, the previous value of `only_explicit` will remain on the field and affect the second dump.

It seems like a lot of complexity has risen out of adding a property to a field that is later inspected by the marshaller to filter fields. I don't think it should be the responsibility of a field to "remember" that it should be dumped. Assigning state info to a persistent field property as a result of a schema operation seems like a poor design choice and `only_explicit` is repeating this pattern.

**Comment by deckar01** [comment]
After looking deeper into it, modifying field instances is not an issue across schema instances, because the `Schema` constructor creates a [deep copy of the `declared_fields` class attribute](https://github.com/marshmallow-code/marshmallow/blob/4c2354c/marshmallow/schema.py#L329) for the instance. It still could be an issue for the field state leaking across multiple `dump()` calls with different arguments on the same schema instance.

**Comment by Archelyst** [comment]
Thanks for analyzing this so thoroughly. I missed the deep copy, will make a test like you suggested as soon as I have some time. 

> It still could be an issue for the field state leaking across multiple dump() calls with different arguments on the same schema instance.

Do you have a specific scenario in mind? I'd include that in the tests then.

@sloria , do we have a decision on the syntax and whether the same should be possible with `exclude`? I'd really like to finish this soon.

**Comment by deckar01** [comment]
> It still could be an issue for the field state leaking across multiple dump() calls with different arguments on the same schema instance.

Since `dump()` doesn't accept an `only` parameter, and field instances aren't shared between schema instances, there is no new field state that is created between `dump()` calls, so it appears that `only_explicit` is unnecessary, an `only` could be applied to the nested field directly.

It is a little misleading that `_update_fields()` is called when the schema is initialized (without a data object), then again in every `dump()` (with the data object it is serializing). Nearly everything that `_update_fields()` does during initialization is repeated unnecessarily during a `dump()` (mainly `__set_field_attrs()`).

That said I think this feature can safely be implemented without modifying `fields.py`. Since we should think of modifications to field instances in `_update_fields()` as initialization logic (which are never changed during a dump), so modifying `only` on a `Nested` field during `_update_fields()` semantically is the same as initializing it with the nested `only` fields supplied to the parent field.

**Comment by Archelyst** [comment]
I agree. The `_update_fields` is a little odd, especially because it's really nothing that should be used form the outside. Well, that's not our concern here (could be addressed for v3) and it seems we can live without `only_explicit`.

**Comment by deckar01** [comment]
Now that I am thinking about it in the context of initialization logic, it would probably be better to strip off the nested field names from `only` and apply them to the nested fields in a private method that is called directly by the `Schema` constructor.  This would avoid introducing additional complexity into the `_update_fields()` method and more explicitly indicate that the `only` parameter is used to override the construction (deep copy) of nested field instances.

**Comment by Archelyst** [comment]
@sloria , let's move on with this. @deckar01 has a [PR with the dot notation](https://github.com/marshmallow-code/marshmallow/pull/468), while I don't agree with the approach (modifying `opts`), it's cleanly implemented and has support for `exclude`. If you wanna go that way, feel free to close this PR and then please merge his. I'd be happy to modify my code here to use the dot-notation and support `exclude` as well. It just would be really nice to come to an end here.

**Comment by sloria** [comment]
I've gone ahead and merged #468 . Thanks so much for your work on this.
